relay-server/src/actors/processor.rs

```diff
@@ -1494,11 +1514,165 @@ impl EnvelopeProcessorService {
+            self.inner.outcome_aggregator.send(TrackOutcome {
+                timestamp,
+                scoping,
+                outcome: Outcome::RateLimited(reason_code),
+                event_id: None,
+                remote_addr: None,
+                category: DataCategory::Profile,
+                quantity: source_quantities.profiles as u32,
+            });
+        }
+    }
+
+    /// Returns `true` if the batches should be rate limited.
+    #[cfg(feature = "processing")]
+    fn rate_limit_batches(
+        &self,
+        cached_rate_limits: RateLimits,
+        scoping: Scoping,
+        bucket_qty: usize,
+        bucket_partitions: &BTreeMap<Option<u64>, Vec<Bucket>>,
+        max_batch_size_bytes: usize,
+        project_state: Arc<ProjectState>,
+    ) -> bool {
+        let mode = {
+            let usage = match project_state.config.transaction_metrics {
+                Some(ErrorBoundary::Ok(ref c)) => c.usage_metric(),
+                _ => false,
+            };
+            ExtractionMode::from_usage(usage)
+        };
+
+        let quotas = &project_state.config.quotas;
+        let item_scoping = ItemScoping {
+            category: DataCategory::Metrics,
+            scoping: &scoping,
+        };
+
+        if let Some(rate_limiter) = self.inner.rate_limiter.as_ref() {
+            // We want to protect the kafka metrics ingestion topics, therefore we ratelimit based on
+            // batches, as the batches are what's sent to kafka.
+            let total_batches: usize = bucket_partitions
+                .values()
+                .map(|buckets| {
+                    // Cheap operation because there's no allocations.
+                    BucketsView::new(buckets)
+                        .by_size(max_batch_size_bytes)
+                        .count()
+                })
+                .sum();
+
+            if cached_rate_limits
+                .check_with_quotas(quotas, item_scoping)
+                .is_limited()
+            {
+                relay_log::info!("dropping {bucket_qty} buckets due to throughput ratelimit");
+                let reason_code = cached_rate_limits
+                    .longest()
+                    .and_then(|limit| limit.reason_code.clone());
+
+                self.drop_buckets_with_outcomes(
+                    reason_code,
+                    total_batches,
+                    scoping,
+                    bucket_partitions,
+                    mode,
+                );
+
+                return true;
+            }
+
+            match rate_limiter.is_rate_limited(quotas, item_scoping, total_batches, false) {
+                Ok(limits) if limits.is_limited() => {
+                    let timestamp = UnixTimestamp::now().as_datetime().unwrap_or_else(Utc::now);
+                    let reason_code = limits.longest().and_then(|limit| limit.reason_code.clone());
+                    self.inner.outcome_aggregator.send(TrackOutcome {
```

Comment:
Doesn't this get tracked twice now, with the call to `drop_buckets_with_outcomes` below?

Also the `reason_code` is calculated twice?

Reply:
ah yeah, missed it while moving code to the function, thanks!